### PR TITLE
Implementing View Search for Pages

### PR DESCRIPTION
**Pull request by rynowak**
/cc @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs [hunk 1]
```diff
@@ -35,18 +36,21 @@ public struct ViewLocationCacheKey : IEquatable<ViewLocationCacheKey>
         /// <param name="viewName">The view name.</param>
         /// <param name="controllerName">The controller name.</param>
         /// <param name="areaName">The area name.</param>
+        /// <param name="pageName">The page name.</param>
         /// <param name="isMainPage">Determines if the page being found is the main page for an action.</param>
         /// <param name="values">Values from <see cref="IViewLocationExpander"/> instances.</param>
         public ViewLocationCacheKey(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs]: It seemed like the right thing to do to make `page` first class.

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs [hunk 2]
```diff
@@ -400,10 +404,24 @@ public string GetAbsolutePath(string executingFilePath, string pagePath)
...
+                viewLocations = _options.ViewLocationFormats;
+            }
+            else if (!string.IsNullOrEmpty(expanderContext.PageName))
+            {
+                viewLocations = _options.PageViewLocationFormats;
+            }
+            else
+            {
+                throw null;
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngine.cs]: I'm not super thrilled about this coupling, but I didn't want to implement a new abstraction.

**Comment by pranavkm** [comment]
Design looks good.

**Comment by rynowak** [comment]
@pranavkm - updated and removed design

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+                        yield return location;
+                    }
+
+                    // For locations with the 'page' token - expand them into an ascending directory search,
+                    // but only up to the pages root.
+                    //
+                    // This is easy because the 'page' token already trims the root directory.
+                    var end = context.PageName.Length;
+
+                    while (end > 0 && (end = context.PageName.LastIndexOf('/', end - 1)) != -1)
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs]: Could you show an example of the values we're producing here?

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorViewEngineTest.cs [hunk 4]
```diff
@@ -1608,6 +1608,108 @@ public void GetNormalizedRouteValue_ConvertsRouteValueToString()
...
+
+            // Act
+            var actual = viewEngine.GetViewLocationFormats(context);
+
+            // Assert
+            Assert.Equal(expected, actual);
+        }
+
+        // This isn't a real case we expect to hit in an app, just making sure we have a reasonable default
+        // for a wierd configuration. In this case we preserve what we did in 1.0.0.
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/RazorViewEngineTest.cs]: weird

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+namespace Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure
+{
+    public class PageViewLocationExpanderTest
+    {
+        [Fact]
+        public void PopulateValues_DoesNothing()
+        {
+            // Arrange
+            var context = CreateContext();
+
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs]: Remove empty line

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+                controllerName: null,
+                areaName: null,
+                pageName: pageName,
+                isMainPage: true)
+            {
+                Values = new Dictionary<string, string>(),
+            };
+        }
+
+        private PageViewLocationExpander CreateViewLocationExpander()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs]: Just inline this

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            var expander = CreateViewLocationExpander();
+
+            // Act
+            var actual = expander.ExpandViewLocations(context, locations);
+
+            // Assert
+            Assert.Equal(expected, actual.ToArray());
+        }
+
+        public void ExpandLocations_ExpandsDirectories_MultipleLocations()
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs]: `[Fact]`

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+            var expander = CreateViewLocationExpander();
+
+            // Act
+            var actual = expander.ExpandViewLocations(context, locations);
+
+            // Assert
+            Assert.Equal(expected, actual.ToArray());
+        }
+
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs]: Can we add a test to verify how this works with the `LanguageViewLocationExpander`? It's the only other one we ship in the box afaik.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs [hunk 9]
```diff
@@ -131,6 +141,7 @@ public override int GetHashCode()
             hashCodeCombiner.Add(ViewName, StringComparer.Ordinal);
             hashCodeCombiner.Add(ControllerName, StringComparer.Ordinal);
             hashCodeCombiner.Add(AreaName, StringComparer.Ordinal);
+            hashCodeCombiner.Add(PageName, StringComparer.Ordinal);
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs]: We should add tests for this type so we know caching works.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs [hunk 10]
```diff
@@ -67,6 +71,11 @@ public struct ViewLocationCacheKey : IEquatable<ViewLocationCacheKey>
         public string AreaName { get; }
 
         /// <summary>
+        /// Gets the page name.
+        /// </summary>
+        public string PageName { get; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs]: Sure

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs]: I added some comments here. I don't think the naming would make it better, its just as ambiguous as before.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/ViewLocationCacheKey.cs]: Discussed, we added a test for the view engine that the caching works equally well for pages.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+                    // For pages, we only handle the 'page' token when it's surrounded by slashes.
+                    //
+                    // Explanation: 
+                    //      We need the ability to 'collapse' the segment which requires us to understand slashes.
+                    //      Imagine a path like /{1}/{0} - we might end up with //{0} if we don't do *something* with
+                    //      the slashes. Instead of picking on (leading or trailing), we choose both. This seems
+                    //      less arbitrary.
+
+
+                    if (!location.Contains("/{1}/"))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs]: `Contains` is more readable and does the correct thing.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/PageViewLocationExpander.cs]: Added

File: test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs (hunk 8 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.RazorPages.Test/Infrastructure/PageViewLocationExpanderTest.cs]: Discussed this. They operate on different parameters.

**Comment by rynowak** [comment]
a8eb5bee7023a2a5d49fcdaea0c25b3416a6921

**Pull request merged**